### ISSUE TITLE
Setter injection results in duplicate shared instances

### ISSUE THREAD

**Issue by beloglazov** [jkolb/FieryCrucible#2]
Continuing from the example in #1, I've tried resolving circular dependencies using setter injection as suggested in the readme. To do that, I've modified the code as follows:

``` Swift
class Test1 {
    init(test2: Test2) {
        println("created Test1")
    }
}

class Test2 {
    var test1: Test1?
    init() {
        println("created Test2")
    }
}

class RootFactory : DependencyFactory {

    func test1() -> Test1 {
        return shared(
            "test1",
            factory: Test1(test2: self.test2()),
            configure: { [unowned self] (instance) in
                println("configure test1")
            }
        )
    }

    func test2() -> Test2 {
        return shared(
            "test2",
            factory: Test2(),
            configure: { [unowned self] (instance) in
                println("configure test2")
                instance.test1 = self.test1()
                println("configured test2")
            }
        )
    }
}

RootFactory().test1()
```

The readme says "This works because all the instances are created in a construction phase before setter injection is triggered allowing references to exist before they are needed." I expected to have a single instance of `Test1` and a single instance of `Test2`, however, the output when I run this example is the following:

```
created Test2
configure test2
created Test1
configure test1
configured test2
created Test1
configure test1
```

In other words, setter injection is triggered before the construction of both instances, and two instances of `Test1` are created: one in the `configure` of `Test2`, and the other in `Test1`'s `factory`. Am I doing something wrong or is there a bug?
